GSdx: Update Post-Processing

Asmodean- opened the pull request:
Updated the GSdx effect suite; improved cel edges of the cel shader; added a debanding shader, and Timothy Lottes' CRT emulation shader, by request.

rz5 commented:
Hell yeah! With this in place, getting the crt-lottes-halation version is basically free

rz5 commented:
The crt-lottes shader looks blurrier than I am expecting. Plus, it does not work on the OpenGL renderers.

Asmodean- commented:
Hmm, It's working fine on OGL, & D3D for me. What GPU are you using? is it AMD?.

Lower FilterCRTAmount to a minus value to sharpen it. It will also depend on output res. Any CRT/scaline filter like this will be pretty varying on emulators, considering the completely arbitrary resolutions.

rz5 commented:
NVIDIA here. GTX 970, latest driver on Windows 10. Only the DX11 renderers show the CRT lottes in action.
Reducing FilterCRTAmount did sharpen it...maybe the default should be a lower value.

The resolution of the masks (compressed TV style, aperature grille, etc) is lower than on RetroArch, even when using something that I assume has a lower internal resolution, like the GBA (mGBA core), with custom viewport height and width set to 960x800 and my display's resolution is 1366x768. Ideally the GSdx CRT lottes implementation would behave a little bit closer to what's available on RetroArch.

Here's a couple of pics for illustration:
http://imgur.com/a/eyyTi

Asmodean- commented:
I've got the same GPU as you, so I've literally no idea why it wouldn't be working for you. I tested in via the gsdx compiler, and fx composer. 

Regarding RetroArch. That doesn't surprize me in the least. RetroArch's post-processing framework is quite a bit more advanced than PCSX2's. 

Saying that - those screenshots would have helped earlier. I can create some workaround variables to try compensate for the lack of functionality. Such as the RetroArch uniforms for input tex size, and video size, etc. These values are taken before scaling is handled. I don't have the free time currently to add this functionality to gsdx.

rz5 commented:
Just enabling just the scanlines pass. Same result for me: works on DX11 renderers but not on OGL ones. Must be something up with my system and not the codebase, so disregard my complaints.

On the topic of the mask's resolution, I notice that if the internal resolution is greater than native, it upscales the resolution of crt-lottes' shadowmasks. Ideally this dependency on internal resolution wouldn't be there and would instead relate to the size of the pcsx2 window or something (however libretro handles it). 
I guess this is what you meant by missing tex/video size uniforms and such.

Asmodean- commented:
That's the best I can do for you. It works fine at 2x scaling or higher. I've replaced some logic to workaround the scaling problems. GSdx doesn't have the functionality of sampling in post-processing before scaling takes places.

Added some extra options to try compensate for this.

turtleli commented on bin/shaders/GSdx.fx:
```diff
@@ -2031,6 +2084,304 @@ float4 BorderPass(float4 colorInput, float2 tex)
+{
+    float2 crtRes = float2(CRTSizeX, CRTSizeY);
+    float2 res = (crtRes * ResolutionScale);
+    pos = floor(pos * res + off) / res;
```
`floor()` seems to shift the image slightly to the bottom right (most visible at native res). Would not rounding here be better?

Asmodean- replied:
Possibly. That's Lottes' logic. Guess he wanted the tc to be less than or equal value. I can't take a look now, I'll fix it tomorrow. I've had that tc offset problem happen before on gsdx with other stuff.

rz5 commented:
GSdx.fx(2302,28): error X3004: undeclared identifier 'CRTSizeX'

ssakash commented:
minor nitpick : (line 26)
> #error GSdx FX is not compatible with GSdx9. Use GSdx OGL or DX10/11

``GSdx9`` could be changed to ``D3D9``
``GSDX OGL`` -> ``OpenGL``
``DX10/11`` -> ``D3D10/11``

No need to mention GSDX multiple times since you have initially mentioned it as ``GSdx FX``

Asmodean- commented:
I actually had it worded like that originally, but people were like "if people are using the wrong backend for post-processing, they might also not know what I mean by d3d" and suggested I change it ;p

So I made it as unambiguous as I could with a single line. But I can change it again. Honestly It doesn't make a difference to me.

ssakash commented:
> I actually had it worded like that originally, but people were like "if people are using the wrong backend for post-processing, they might also not know what I mean by d3d" and suggested I change it ;p

that kinda seems reasonable :p , it's your decision though :)

Asmodean- commented:
Alright, that's updated. I've changed a few more things to try compensate for gs scaling issues. I've also added an image resolution scale, and mask resolution scale. Which alter the image res, and mask res _somewhat_independently of the RT size. For easier adjusting.

I don't like simply adding more options for X, Y, etc, tbh. But I'd to add quite a few due to scaling problems. So play around with them, if you're hot happy with the defaults.

This is finished, unless there are any more issues.
